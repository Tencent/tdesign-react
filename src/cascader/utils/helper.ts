import { TreeNode, CascaderContextType, TreeNodeValue, TreeOptionData } from '../interface';

/**
 * 获取node的全部路径
 * @param node
 * @returns
 */
export function getFullPathLabel(node: TreeNode) {
  return node
    .getPath()
    .map((node: TreeNode) => node.label)
    .join('/');
}

/**
 * treeValue计算方法
 * @param value
 * @returns
 */
export const getTreeValue = (value: CascaderContextType['value']) => {
  let treeValue: TreeNodeValue[] = [];
  if (Array.isArray(value)) {
    if (value.length > 0 && typeof value[0] === 'object') {
      treeValue = (value as TreeOptionData[]).map((val) => val.value);
    } else if (value.length) {
      treeValue = value as TreeNodeValue[];
    }
  } else if (value) {
    if (typeof value === 'object') {
      treeValue = [(value as TreeOptionData).value];
    } else {
      treeValue = [value];
    }
  }
  return treeValue;
};

/**
 * 判断node是否在页面上
 * @param node Node
 */
export function isInPage(node: Node) {
  return node === document.body ? false : document.body.contains(node);
}

export default {
  getFullPathLabel,
};
