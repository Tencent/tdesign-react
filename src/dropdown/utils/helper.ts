import React from 'react';
import { DropdownOption } from '../type';
import DropdownMenu from '../DropdownMenu';

export const getOptionsFromChildren = (children: React.ReactElement): DropdownOption[] => {
  if (children.type === DropdownMenu) {
    const groupChildren = children.props.children as React.ReactElement;
    if (Array.isArray(groupChildren)) {
      return getOptionsFromChildren(groupChildren);
    }
  }

  if (Array.isArray(children)) {
    return children.map((item: React.ReactElement) => {
      const groupChildren = item.props.children;
      if (Array.isArray(groupChildren)) {
        const contentIdx = groupChildren.findIndex((v) => typeof v === 'string');
        return {
          ...item.props,
          content: groupChildren[contentIdx],
          children: getOptionsFromChildren(groupChildren[1]),
        };
      }
      return { ...item.props, content: item.props.children || item.props.content };
    });
  }

  return [];
};
