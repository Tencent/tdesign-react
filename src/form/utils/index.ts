import type { NamePath } from '../type';

// 获取 formMap 管理的数据
export function getMapValue(name: NamePath, formMapRef: React.MutableRefObject<Map<any, any>>) {
  if (!formMapRef.current) return;

  // 提取所有 map key
  const mapKeys = [...formMapRef.current.keys()];
  // 转译为字符串后比对 key 兼容数组格式
  const key = mapKeys.find((key) => String(key) === String(name));
  // 拿到 key 引用地址获取 value
  return formMapRef.current.get(key);
}

// 将数据整理成对象，数组 name 转化嵌套对象: ['user', 'name'] => { user: { name: '' } }
export function calcFieldValue(name: NamePath, value: any) {
  if (Array.isArray(name)) {
    const fieldValue = name.reduceRight((prev, curr) => {
      const arr = [];
      if (/^\d+$/.test(String(curr))) arr[curr] = prev;
      return arr.length ? arr : { [curr]: prev };
    }, value);
    return { ...fieldValue };
  }

  return { [name]: value };
}

// 通过对象数据类型获取 map 引用: { user: { name: '' } } => formMap.get(['user', 'name'])
export function travelMapFromObject(
  obj: Record<any, any>,
  formMapRef: React.MutableRefObject<Map<any, any>>,
  callback: Function,
) {
  for (const [mapName, formItemRef] of formMapRef.current.entries()) {
    // 支持嵌套数据结构
    if (Array.isArray(mapName)) {
      // 创建唯一临时变量 symbol
      const symbol = Symbol('name');
      let fieldValue = null;

      for (let i = 0; i < mapName.length; i++) {
        const item = mapName[i];
        if (Reflect.has(fieldValue || obj, item)) {
          fieldValue = Reflect.get(fieldValue || obj, item);
        } else {
          // 当反射无法获取到值则重置为 symbol
          fieldValue = symbol;
          break;
        }
      }

      // 非 symbol 说明获取到了值
      if (fieldValue !== symbol) {
        callback(formItemRef, fieldValue);
      }
    } else if (Reflect.has(obj, mapName)) {
      callback(formItemRef, obj[mapName]);
    }
  }
}
