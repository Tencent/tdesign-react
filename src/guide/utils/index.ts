import { scrollToParentVisibleArea } from './getScrollParent';
import getRelativePosition from './getRelativePosition';
import getTargetElm from './getTargetElm';
import scrollToElm from './scrollToElm';
import useMount from './useMount';

export { scrollToParentVisibleArea, getRelativePosition, getTargetElm, scrollToElm, useMount };
