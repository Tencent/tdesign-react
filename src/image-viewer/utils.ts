export const downloadFile = function (imgSrc) {
  const image = new Image();
  const name = imgSrc?.split?.('/').pop() || Math.random().toString(32).slice(2);

  image.setAttribute('crossOrigin', 'anonymous');

  image.onload = () => {
    const canvas = document.createElement('canvas');
    canvas.width = image.width;
    canvas.height = image.height;

    const context = canvas.getContext('2d');
    context.drawImage(image, 0, 0, image.width, image.height);
    canvas.toBlob((blob) => {
      const url = URL.createObjectURL(blob);
      const a = document.createElement('a');
      a.download = name;
      a.href = url;
      a.click();
      a.remove();
      URL.revokeObjectURL(url);
    });
  };
  image.src = imgSrc;
};
