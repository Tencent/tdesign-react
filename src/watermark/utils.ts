const toLowercaseSeparator = (key: string) => key.replace(/([A-Z])/g, '-$1').toLowerCase();

// style对象转字符串
export const getStyleStr = (style: React.CSSProperties): string =>
  Object.keys(style)
    .map((key: keyof React.CSSProperties) => `${toLowercaseSeparator(key)}: ${style[key]};`)
    .join(' ');
