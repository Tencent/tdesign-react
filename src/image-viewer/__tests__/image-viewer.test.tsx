import React from 'react';
import { render, act, fireEvent, waitFor, vi, mockTimeout } from '@test/utils';
import userEvent from '@testing-library/user-event';

import { ImageViewer } from '../index';

const imgUrl = 'https://tdesign.gtimg.com/demo/demo-image-1.png';
const imgUrl2 = 'https://tdesign.gtimg.com/demo/demo-image-2.png';
// const errorImgUrl = 'https://tdesixxxxxxxxgn.gtimg.com/demo/demo-image-1.png';

describe('ImageViewer', () => {
  const triggerText = '预览单张图片';

  test('base', async () => {
    const onClose = vi.fn();
    const BasicImageViewer = () => {
      const trigger = ({ onOpen }) => <span onClick={onOpen}>{triggerText}</span>;
      return <ImageViewer trigger={trigger} images={[imgUrl]} onClose={onClose} />;
    };
    const { getByText } = render(<BasicImageViewer />);

    // 点击前，没有元素存在
    const imgContainer = document.querySelector('.t-image-viewer-preview-image');
    expect(imgContainer).toBeNull();

    // 模拟鼠标点击
    act(() => {
      fireEvent.click(getByText(triggerText));
    });

    // 鼠标点击后，有元素
    expect(onClose).toHaveBeenCalledTimes(1);
    const imgModal = document.querySelector('.t-image-viewer__modal-pic');
    expect(imgModal).toBeTruthy();

    // 模拟鼠标点击关闭
    const closeBtn = document.querySelector('.t-image-viewer__modal-close-bt');
    act(() => {
      fireEvent.click(closeBtn);
    });
    // 点击后，没有元素存在
    expect(onClose).toHaveBeenCalledTimes(2);
    await mockTimeout(() => expect(document.querySelector('.t-image-viewer-preview-image')).toBeNull());
  });

  test('base:trigger is not Fn', async () => {
    const BasicImageViewer = () => {
      const trigger = <span>{triggerText}</span>;
      return <ImageViewer trigger={trigger} />;
    };
    const { getByText } = render(<BasicImageViewer />);
    expect(getByText(triggerText)).toBeTruthy();
  });
});

describe('ImageViewerMini', () => {
  const triggerText = '预览单张图片';

  test('modeless', async () => {
    const onClose = vi.fn();
    const BasicImageViewer = () => {
      const trigger = ({ onOpen }) => <span onClick={onOpen}>{triggerText}</span>;
      return <ImageViewer trigger={trigger} images={[imgUrl]} onClose={onClose} mode="modeless" />;
    };
    const { getByText } = render(<BasicImageViewer />);

    // 模拟鼠标点击
    act(() => {
      fireEvent.click(getByText(triggerText));
    });

    // 鼠标点击后，有 mini 元素
    const miniFooter = await waitFor(() => document.querySelector('.t-image-viewer-mini__footer'));
    expect(miniFooter).toBeTruthy();

    // 模拟鼠标点击关闭
    const closeBtn = await waitFor(() => document.querySelector('.t-icon-close'));
    act(() => {
      fireEvent.click(closeBtn);
    });
    expect(onClose).toHaveBeenCalledTimes(2);
  });
});

describe('ImageViewerModel', () => {
  const triggerText = '预览单张图片';
  test('base', async () => {
    const user = userEvent.setup();
    const onClose = vi.fn();
    const onIndexChange = vi.fn();
    const BasicImageViewer = () => {
      const trigger = ({ onOpen }) => <span onClick={onOpen}>{triggerText}</span>;
      return (
        <ImageViewer
          trigger={trigger}
          images={[imgUrl, imgUrl2]}
          onClose={onClose}
          onIndexChange={onIndexChange}
          closeOnOverlay={true}
        />
      );
    };
    const { getByText } = render(<BasicImageViewer />);

    // 模拟鼠标点击
    act(() => {
      fireEvent.click(getByText(triggerText));
    });

    // 模拟键盘事件
    await user.type(document.body, '{Escape}');
    expect(onClose).toHaveBeenCalledTimes(2);

    await user.type(document.body, '{ArrowRight}');
    expect(onIndexChange).toHaveBeenCalledTimes(1);

    await user.type(document.body, '{ArrowLeft}');
    expect(onIndexChange).toHaveBeenCalledTimes(2);

    // 鼠标点击遮罩
    const mask = await waitFor(() => document.querySelector('.t-image-viewer__modal-mask'));
    act(() => {
      fireEvent.click(mask);
    });
    expect(onClose).toHaveBeenCalledTimes(3);
  });

  test('single', async () => {
    const user = userEvent.setup();
    const BasicImageViewer = () => {
      const trigger = ({ onOpen }) => <span onClick={onOpen}>{triggerText}</span>;
      return <ImageViewer trigger={trigger} images={[imgUrl]} />;
    };
    const { getByText } = render(<BasicImageViewer />);

    // 模拟鼠标点击
    act(() => {
      fireEvent.click(getByText(triggerText));
    });

    const img = document.querySelector('.t-image-viewer__modal-image');
    expect(getComputedStyle(img).transform).toBe('rotateZ(0deg) scale(1)');

    await user.type(document.body, '{ArrowUp}');
    expect(getComputedStyle(img).transform).toBe('rotateZ(0deg) scale(1.5)');

    await user.type(document.body, '{ArrowDown}');
    expect(getComputedStyle(img).transform).toBe('rotateZ(0deg) scale(1)');

    const rotateIcon = await waitFor(() => document.querySelector('.t-icon-rotation'));
    // 模拟鼠标点击
    act(() => {
      fireEvent.click(rotateIcon);
    });
    expect(getComputedStyle(img).transform).toBe('rotateZ(-90deg) scale(1)');

    const resetIcon = await waitFor(() => document.querySelector('.t-icon-image'));
    // 模拟鼠标点击
    act(() => {
      fireEvent.click(resetIcon);
    });
    expect(getComputedStyle(img).transform).toBe('rotateZ(0deg) scale(1)');
  });

  test('closeBtn', async () => {
    const BasicImageViewer = () => {
      const trigger = ({ onOpen }) => <span onClick={onOpen}>{triggerText}</span>;
      return <ImageViewer trigger={trigger} images={[imgUrl]} closeBtn={() => <span>closeBtn</span>} />;
    };
    const { getByText } = render(<BasicImageViewer />);

    // 模拟鼠标点击
    act(() => {
      fireEvent.click(getByText(triggerText));
    });
    expect(getByText('closeBtn')).toBeTruthy();
  });
});
