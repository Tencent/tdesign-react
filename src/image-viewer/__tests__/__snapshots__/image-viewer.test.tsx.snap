// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`album.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-image-viewer__base"
    style="width: 240px; height: 240px;"
  >
    <div
      class="tdesign-demo-image-viewer__ui-image"
    >
      <img
        alt="test"
        class="tdesign-demo-image-viewer__ui-image--img"
        src="https://tdesign.gtimg.com/demo/demo-image-1.png"
      />
      <div
        class="tdesign-demo-image-viewer__ui-image--hover"
      >
        <span>
          <svg
            class="t-icon t-icon-browse"
            fill="none"
            height="1em"
            name="browse"
            style="font-size: 1.4em;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
              fill="currentColor"
              fill-opacity="0.9"
            />
            <path
              d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
           预览
        </span>
      </div>
      <div
        class="tdesign-demo-image-viewer__ui-image--footer"
      >
        <span
          class="tdesign-demo-image-viewer__ui-image--title"
        >
          相册封面标题
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`albumIcons.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-image-viewer__base"
    style="width: 240px; height: 240px;"
  >
    <div
      class="tdesign-demo-image-viewer__ui-image"
    >
      <img
        alt="test"
        class="tdesign-demo-image-viewer__ui-image--img"
        src="https://tdesign.gtimg.com/demo/demo-image-1.png"
      />
      <div
        class="tdesign-demo-image-viewer__ui-image--hover"
      >
        <span>
          <svg
            class="t-icon t-icon-browse"
            fill="none"
            height="1em"
            name="browse"
            style="font-size: 1.4em;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
              fill="currentColor"
              fill-opacity="0.9"
            />
            <path
              d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
           预览
        </span>
      </div>
      <div
        class="tdesign-demo-image-viewer__ui-image--footer"
      >
        <span
          class="tdesign-demo-image-viewer__ui-image--title"
        >
          相册封面标题
        </span>
        <span
          class="tdesign-demo-image-viewer__ui-image--icons"
        >
          <svg
            class="t-icon t-icon-browse"
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
              fill="currentColor"
              fill-opacity="0.9"
            />
            <path
              d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
          <div
            class="t-popup__reference"
          >
            <svg
              class="t-icon t-icon-ellipsis"
              classname="tdesign-demo-image-viewer__ui-image--ellipsis"
              fill="none"
              height="1em"
              viewBox="0 0 16 16"
              width="1em"
            >
              <path
                d="M3 9a1 1 0 110-2 1 1 0 010 2zM7 8a1 1 0 102 0 1 1 0 00-2 0zM12 8a1 1 0 102 0 1 1 0 00-2 0z"
                fill="currentColor"
                fill-opacity="0.9"
              />
            </svg>
          </div>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`base.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-image-viewer__base"
  >
    <div
      class="tdesign-demo-image-viewer__ui-image"
    >
      <img
        alt="test"
        class="tdesign-demo-image-viewer__ui-image--img"
        src="https://tdesign.gtimg.com/demo/demo-image-1.png"
      />
      <div
        class="tdesign-demo-image-viewer__ui-image--hover"
      >
        <span>
          <svg
            class="t-icon t-icon-browse"
            fill="none"
            height="1em"
            name="browse"
            style="font-size: 1.4em;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
              fill="currentColor"
              fill-opacity="0.9"
            />
            <path
              d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
           预览
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`block.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <div
      class="tdesign-demo-image-viewer__ui-image"
    >
      <img
        alt="test"
        class="tdesign-demo-image-viewer__ui-image--img"
        src="https://tdesign.gtimg.com/demo/demo-image-1.png"
      />
      <div
        class="tdesign-demo-image-viewer__ui-image--hover"
      >
        <span>
          <svg
            class="t-icon t-icon-browse"
            fill="none"
            height="1em"
            name="browse"
            style="font-size: 1.4em;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
              fill="currentColor"
              fill-opacity="0.9"
            />
            <path
              d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
           预览
        </span>
      </div>
    </div>
  </div>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <div
      class="tdesign-demo-image-viewer__ui-image"
    >
      <img
        alt="test"
        class="tdesign-demo-image-viewer__ui-image--img"
        src="https://tdesign.gtimg.com/demo/demo-image-1.png"
      />
      <div
        class="tdesign-demo-image-viewer__ui-image--hover"
      >
        <span>
          <svg
            class="t-icon t-icon-browse"
            fill="none"
            height="1em"
            name="browse"
            style="font-size: 1.4em;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
              fill="currentColor"
              fill-opacity="0.9"
            />
            <path
              d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
           预览
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`button.jsx 1`] = `
<DocumentFragment>
  <button
    class="t-button t-button--theme-primary t-button--variant-base"
    type="button"
  >
    <span
      class="t-button__text"
    >
      预览单张图片
    </span>
  </button>
</DocumentFragment>
`;

exports[`error.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <img
      alt="test"
      class="tdesign-demo-image-viewer__ui-image--img"
      src="https://tdesign.gtimg.com/demo/demo-image-1.png"
    />
    <div
      class="tdesign-demo-image-viewer__ui-image--hover"
    >
      <span>
        <svg
          class="t-icon t-icon-browse"
          fill="none"
          height="1em"
          style="font-size: 1.4em;"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
         预览
      </span>
    </div>
  </div>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <img
      alt="test"
      class="tdesign-demo-image-viewer__ui-image--img"
      src="https://tdesign.gtimg.com/demo/demo-image-error1.png"
    />
    <div
      class="tdesign-demo-image-viewer__ui-image--hover"
    >
      <span>
        <svg
          class="t-icon t-icon-browse"
          fill="none"
          height="1em"
          style="font-size: 1.4em;"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
         预览
      </span>
    </div>
  </div>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <img
      alt="test"
      class="tdesign-demo-image-viewer__ui-image--img"
      src="https://tdesign.gtimg.com/demo/demo-image-error2.png"
    />
    <div
      class="tdesign-demo-image-viewer__ui-image--hover"
    >
      <span>
        <svg
          class="t-icon t-icon-browse"
          fill="none"
          height="1em"
          style="font-size: 1.4em;"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
         预览
      </span>
    </div>
  </div>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <img
      alt="test"
      class="tdesign-demo-image-viewer__ui-image--img"
      src="https://tdesign.gtimg.com/demo/demo-image-error3.png"
    />
    <div
      class="tdesign-demo-image-viewer__ui-image--hover"
    >
      <span>
        <svg
          class="t-icon t-icon-browse"
          fill="none"
          height="1em"
          style="font-size: 1.4em;"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
         预览
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`modeless.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-image-viewer__base"
  >
    <div
      class="tdesign-demo-image-viewer__ui-image"
    >
      <img
        alt="test"
        class="tdesign-demo-image-viewer__ui-image--img"
        src="https://tdesign.gtimg.com/demo/demo-image-1.png"
      />
      <div
        class="tdesign-demo-image-viewer__ui-image--hover"
      >
        <span>
          <svg
            class="t-icon t-icon-browse"
            fill="none"
            height="1em"
            style="font-size: 1.4em;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
              fill="currentColor"
              fill-opacity="0.9"
            />
            <path
              d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
           预览
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`multiple.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <img
      alt="test"
      class="tdesign-demo-image-viewer__ui-image--img"
      src="https://tdesign.gtimg.com/demo/demo-image-1.png"
    />
    <div
      class="tdesign-demo-image-viewer__ui-image--hover"
    >
      <span>
        <svg
          class="t-icon t-icon-browse"
          fill="none"
          height="1em"
          style="font-size: 1.4em;"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
         预览
      </span>
    </div>
  </div>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <img
      alt="test"
      class="tdesign-demo-image-viewer__ui-image--img"
      src="https://tdesign.gtimg.com/demo/demo-image-3.png"
    />
    <div
      class="tdesign-demo-image-viewer__ui-image--hover"
    >
      <span>
        <svg
          class="t-icon t-icon-browse"
          fill="none"
          height="1em"
          style="font-size: 1.4em;"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
         预览
      </span>
    </div>
  </div>
  <div
    class="tdesign-demo-image-viewer__ui-image tdesign-demo-image-viewer__base"
  >
    <img
      alt="test"
      class="tdesign-demo-image-viewer__ui-image--img"
      src="https://tdesign.gtimg.com/demo/demo-image-2.png"
    />
    <div
      class="tdesign-demo-image-viewer__ui-image--hover"
    >
      <span>
        <svg
          class="t-icon t-icon-browse"
          fill="none"
          height="1em"
          style="font-size: 1.4em;"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M10.88 8a2.88 2.88 0 11-5.76 0 2.88 2.88 0 015.76 0zm-1 0a1.88 1.88 0 10-3.76 0 1.88 1.88 0 003.76 0z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M1.12 8.23A7.72 7.72 0 008 12.5c2.9 0 5.54-1.63 6.88-4.27L15 8l-.12-.23A7.73 7.73 0 008 3.5a7.74 7.74 0 00-6.88 4.27L1 8l.12.23zM8 11.5A6.73 6.73 0 012.11 8 6.73 6.73 0 0113.9 8 6.74 6.74 0 018 11.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
         预览
      </span>
    </div>
  </div>
</DocumentFragment>
`;
